zerver/lib/narrow.py

```diff
@@ -333,8 +335,24 @@ def add_term(self, query: Select, term: Dict[str, Any]) -> Select:
         return method(query, operand, maybe_negate)
 
     def by_has(self, query: Select, operand: str, maybe_negate: ConditionTransform) -> Select:
-        if operand not in ["attachment", "image", "link"]:
+        if operand not in ["attachment", "image", "link", "reaction"]:
             raise BadNarrowOperatorError("unknown 'has' operand " + operand)
+        if operand == "reaction":
+            subquery = (
+                select(column("message_id", Integer).label("reaction_message_id"))
+                .select_from(table("zerver_reaction"))
+                .alias("zerver_reaction_subtable")
+            )
+            if self.need_user_message:
+                return query.join(
+                   subquery,
+                    self.msg_id_column == literal_column("zerver_reaction_subtable.reaction_message_id", Integer),
+                ).group_by(self.msg_id_column, literal_column("zerver_usermessage.flags", Integer))
+            else:
+                return query.join(
+                    subquery,
+                    self.msg_id_column == literal_column("zerver_reaction_subtable.reaction_message_id", Integer),
+                ).group_by(self.msg_id_column)
```

Comment:
We should do a fancier check that just does an `exists` check, since that's all we care about here. Might make this a bit more efficient.